### ISSUE TITLE
RuntimeError: parse error

### ISSUE THREAD

**Issue by curbsaleem** [rubysolo/dentaku#62]
RuntimeError: parse error at: '#met_values#exercise - #food_frequency#vegall )*2 - 11 + (#exercise_days *2) '
dentaku-2.0.4/lib/dentaku/tokenizer.rb:16:in `tokenize'

My formula was "(#met_values - #food_frequency )*2 - 11 + (#exercise_days *2) " 
------------------------
  def validate_formula
    calculator = Dentaku::Calculator.new
    formula.scan(/#\S+/).each do |hash|
      calculator.store(hash: 1)
    end
    errors.add(:formula, "is not valid") unless calculator.evaluate(formula)
  end

Plz let me know if i m missing any thing

**Comment by Fuelen** [comment]
remove # from your formula
```ruby
formula = '(#met_values - #food_frequency )*2 - 11 + (#exercise_days *2)'
formula.scan(/#\S+/).each do |hash|
  calculator.store(hash => 1)
end
errors.add(:formula, "is not valid") unless calculator.evaluate(formula.gsub('#', ''))
```
or
```ruby
formula = '(met_values - food_frequency )*2 - 11 + (exercise_days *2)'
variables = formula.split(/[0-9]*[^a-z_\d]+/).delete_if { |var| var.blank? }
variables.each do |variable|
  calculator.store(variable => 1)
end
errors.add(:formula, "is not valid") unless calculator.evaluate(formula)
```
But this both variants will raise an exception if syntax errors are present or function isn't exists.
So i recommend to use this code
```ruby
formula = '(met_values - food_frequency )*2 - 11 + (exercise_days *2)'
begin
  calculator.evaluate(formula)
rescue RuntimeError => e
  errors.add(:text, "has #{e.message.downcase}")
rescue Exception
  errors.add(:formula, "has syntax errors")
end
```

**Comment by curbsaleem** [comment]
Is thr any way to parse  reference variable starting from # ?
Thanks for your reply!

**Comment by Fuelen** [comment]
See my first example.

**Comment by curbsaleem** [comment]
that is giving error!. may be i need to store 
calculator.store(hash.gsub('#', '') => 1)

**Comment by Fuelen** [comment]
Yes, you are right.

**Comment by curbsaleem** [comment]
basically i want to know how i can parse  reference variable which starts from any special character

**Comment by Fuelen** [comment]
Just replace `#` to any another character (or write your pattern like this /([#@!]\S+/ ), but not to use built-in operators. Or you need to parse all references, that are begins from built-in operators?

**Comment by rubysolo** [comment]
I just pushed an update to master that makes it easier to customize tokenization, so you should be able to do something like this:

```ruby
> custom = Dentaku::TokenScanner.new(:identifier, '#\w+\b', ->(raw) { raw.strip.downcase })
=> #<Dentaku::TokenScanner ... >
> Dentaku::TokenScanner.register_scanner(:custom, custom)
=> #<Dentaku::TokenScanner ... >
> formula = '#exercise_days * 2'
=> "#exercise_days * 2"
> Dentaku(formula, '#exercise_days' => 5)
=> 10
```

I'm not sure what your requirements are for variable names -- the `#\w+\b` regex will match word characters with a `#` at the beginning.  If you need something different, you could change the regex.

Hope this helps!

**Comment by rubysolo** [comment]
This is now released as [v2.0.5](https://github.com/rubysolo/dentaku/commit/176ad5e195662d98e1044e4c584c18ca9c465a9f)
